sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala

```diff
@@ -1108,6 +1125,366 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+        "   +83028485        ",
+      (Decimal(0),
+        "9999999999999999.999999999999999S") ->
+        "               0+                ",
+      (Decimal(unscaled = 43100000000L, precision = 38, scale = 10),
+        "9999999999999999.999999999999999S") ->
+        "               4.31+             ",
+      (Decimal(-454.8),
+        "99G999.9S") ->
+        "   454.8-",
+      (Decimal(-454.8),
+        "00G000.0S") ->
+        "00,454.8-",
+      (Decimal(-454),
+        "S999") ->
+        "-454",
+      (Decimal(-454),
+        "999S") ->
+        "454-",
+      (Decimal(-12454.8),
+        "99G999D9S") ->
+        "12,454.8-",
+      (Decimal(-83028485),
+        "99999999999.9999999S") ->
+        "   83028485-        "
+    ).foreach { case ((decimal, format), expected) =>
+      var expr: Expression = ToCharacter(Literal(decimal), Literal(format))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+
+      expr = TryToCharacter(Literal(decimal), Literal(format))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+    }
+
+    // Test 'MI'
+    Seq(
+      (Decimal(4.31),
+        "9999999999999999.999999999999999MI") ->
+        "               4.31               ",
+      (Decimal(0),
+        "9999999999999999.999999999999999MI") ->
+        "               0                  ",
+      (Decimal(unscaled = 43100000000L, precision = 38, scale = 10),
+        "9999999999999999.999999999999999MI") ->
+        "               4.31               ",
+      (Decimal(-454.8),
+        "99G999.9MI") ->
+        "   454.8- ",
+      (Decimal(-454.8),
+        "00G000.0MI") ->
+        "00,454.8- ",
+      (Decimal(-454),
+        "999MI") ->
+        "454- ",
+      (Decimal(-12454.8),
+        "99G999D9MI") ->
+        "12,454.8- "
+    ).foreach { case ((decimal, format), expected) =>
+      var expr: Expression = ToCharacter(Literal(decimal), Literal(format))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+
+      expr = TryToCharacter(Literal(decimal), Literal(format))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+    }
+
+    // Test 'PR'
+    Seq(
+      (Decimal(4.31),
+        "9999999999999999.999999999999999PR") ->
+        "               4.31               ",
+      (Decimal(0),
+        "9999999999999999.999999999999999PR") ->
+        "               0                  ",
+      (Decimal(unscaled = 43100000000L, precision = 38, scale = 10),
+        "9999999999999999.999999999999999PR") ->
+        "               4.31               ",
+      (Decimal(-123),
+        "9999999999999999.999PR") ->
+        "             <123>    ",
+      (Decimal(-123.4),
+        "9999999999999999.999PR") ->
+        "             <123.4>  ",
```

Comment:
nit: would it be nicer to remove one space from the left and add one space to the right?

Reply:
We could do that, but then the positioning of the decimal point would be inconsistent between positive and negative numbers. Consider the following examples:

```
select to_char(Decimal(8), '9PR')
> ' 8 '
select to_char(Decimal(-8), '9PR')
> '<8>'
select to_char(Decimal(8.8), '9.9PR')
> ' 8.8 '
select to_char(Decimal(-8.8), '9.9PR')
> '<8.8>'
```

In these cases the result string gains two extra characters from the `PR` which can be either spaces or `<`/`>`. Thoughts?